### ISSUE TITLE
Lint `clippy::new_without_default` failed to modify the code correctly when the `new()` method is attributed with `#[new]` of PyO3

### ISSUE THREAD

**Issue by Xuanmizhen** [rust-lang/rust-clippy#17361]
### Summary

MVE (with PyO3 0.28.3, Rust 2024):
```rust
use pyo3::prelude::*;

#[pyclass]
pub struct A {}

#[pymethods]
impl A {
    #[new]
    pub fn new() -> Self {
        Self {}
    }
}
```
The code compiles, but clippy suggested:
```
 % cargo clippy
warning: you should consider adding a `Default` implementation for `A`
  --> src/lib.rs:9:5
   |
 9 | /     pub fn new() -> Self {
10 | |         Self {}
11 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.96.0/index.html#new_without_default
   = note: `#[warn(clippy::new_without_default)]` on by default
help: try adding this
   |
 7 + impl Default for A {
 8 +     fn default() -> Self {
 9 +         Self::new()
10 +     }
11 + }
   |

warning: `rsbugdemo` (lib) generated 1 warning (run `cargo clippy --fix --lib -p rsbugdemo -- ` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.02s
```
But when running `cargo clippy --fix --lib -p rsbugdemo -- `:
```
 % cargo clippy --fix --lib -p rsbugdemo -- 
    Checking rsbugdemo v0.1.0 (/Users/dengshuai/GitHub/Xuanmizhen/rsbugdemo)
error: errors present after applying fixes to crate `rsbugdemo`
 --> src/lib.rs
  = cause: error: #[pymethods] cannot be used on trait impl blocks
            --> src/lib.rs:7:6
             |
           7 | impl Default for A {
             |      ^^^^^^^
           
           
           error: cannot find attribute `pyo3` in this scope
            --> src/lib.rs:6:1
             |
           6 | #[pymethods]
             | ^^^^^^^^^^^^
             |
             = note: `pyo3` is in scope, but it is a crate, not an attribute
             = note: this error originates in the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)
           help: `pyo3` is an attribute that can be used by the derive macros `FromPyObject`, `IntoPyObject` and `IntoPyObjectRef`, you might be missing a `derive` attribute
             |
           7 + #[derive(FromPyObject, IntoPyObject, IntoPyObjectRef)]
           8 | impl Default for A {
             |
           
           
           error: cannot find attribute `new` in this scope
             --> src/lib.rs:14:7
              |
           14 |     #[new]
              |       ^^^
           
           
           error: aborting due to 3 previous errors
help: to report this as a bug, open an issue at https://github.com/rust-lang/rust-clippy/issues, quoting the full output of this command
help: to possibly apply more fixes, pass in the `--broken-code` flag
note: original diagnostics will follow:
warning: you should consider adding a `Default` implementation for `A`
  --> src/lib.rs:9:5
   |
 9 | /     pub fn new() -> Self {
10 | |         Self {}
11 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.96.0/index.html#new_without_default
   = note: `#[warn(clippy::new_without_default)]` on by default
help: try adding this
   |
 7 + impl Default for A {
 8 +     fn default() -> Self {
 9 +         Self::new()
10 +     }
11 + }
   |

warning: `rsbugdemo` (lib) generated 1 warning (run `cargo clippy --fix --lib -p rsbugdemo -- --no-deps` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.11s
```
Is this a bug of Clippy? Clippy cannot help me create the correct code directly in this case, and it suggests that I can open an issue here. Maybe Clippy should deal with code in some correct way, or maybe Clippy should not consider modifying this code.
The MVE is also put here: https://github.com/Xuanmizhen/rsbugdemo
So you can clone it and have a try.

### Version

```text
rustc 1.96.1 (31fca3adb 2026-06-26)
binary: rustc
commit-hash: 31fca3adb283cc9dfd56b49cdee9a96eb9c96ffd
commit-date: 2026-06-26
host: aarch64-apple-darwin
release: 1.96.1
LLVM version: 22.1.2
```

### Error output

<details><summary>Backtrace</summary>
  <p>

  ```
 % RUST_BACKTRACE=1 cargo clippy --fix --lib -p rsbugdemo --
    Checking rsbugdemo v0.1.0 (/Users/dengshuai/GitHub/Xuanmizhen/rsbugdemo)
error: errors present after applying fixes to crate `rsbugdemo`
 --> src/lib.rs
  = cause: error: #[pymethods] cannot be used on trait impl blocks
            --> src/lib.rs:7:6
             |
           7 | impl Default for A {
             |      ^^^^^^^
           
           
           error: cannot find attribute `pyo3` in this scope
            --> src/lib.rs:6:1
             |
           6 | #[pymethods]
             | ^^^^^^^^^^^^
             |
             = note: `pyo3` is in scope, but it is a crate, not an attribute
             = note: this error originates in the attribute macro `pymethods` (in Nightly builds, run with -Z macro-backtrace for more info)
           help: `pyo3` is an attribute that can be used by the derive macros `FromPyObject`, `IntoPyObject` and `IntoPyObjectRef`, you might be missing a `derive` attribute
             |
           7 + #[derive(FromPyObject, IntoPyObject, IntoPyObjectRef)]
           8 | impl Default for A {
             |
           
           
           error: cannot find attribute `new` in this scope
             --> src/lib.rs:14:7
              |
           14 |     #[new]
              |       ^^^
           
           
           error: aborting due to 3 previous errors
help: to report this as a bug, open an issue at https://github.com/rust-lang/rust-clippy/issues, quoting the full output of this command
help: to possibly apply more fixes, pass in the `--broken-code` flag
note: original diagnostics will follow:
warning: you should consider adding a `Default` implementation for `A`
  --> src/lib.rs:9:5
   |
 9 | /     pub fn new() -> Self {
10 | |         Self {}
11 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.96.0/index.html#new_without_default
   = note: `#[warn(clippy::new_without_default)]` on by default
help: try adding this
   |
 7 + impl Default for A {
 8 +     fn default() -> Self {
 9 +         Self::new()
10 +     }
11 + }
   |

warning: `rsbugdemo` (lib) generated 1 warning (run `cargo clippy --fix --lib -p rsbugdemo -- --no-deps` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.33s
  ```

  </p>
</details>
